dask/backends.py

```diff
@@ -0,0 +1,118 @@
+
+from dask import config
+from dask.compatibility import entry_points
+
+
+class DaskBackendEntrypoint(ABC):
+    """Base Collection-Backend Entrypoint Class
+
+    Most methods in this class correspond to collection-creation
+    for a specific library backend. Once a collection is created,
+    the existing data will be used to dispatch compute operations
+    within individual tasks. The backend is responsible for
+    ensuring that these data-directed dispatch functions are
+    registered when ``__init__`` is called.
+    """
+
+    @abstractmethod
+    def __init__(self):
+        """Register data-directed dispatch functions
+
+        This may be a ``pass`` operation if the data dispatch functions
+        are already registered within the same module that the
+        ``DaskBackendEntrypoint`` subclass is defined.
+        """
+        raise NotImplementedError
+
+
+@lru_cache(maxsize=1)
+def detect_entrypoints(entry_point_name):
+    entrypoints = entry_points(entry_point_name)
+    return {ep.name: ep for ep in entrypoints}
+
+
+BackendEntrypointType = TypeVar(
+    "BackendEntrypointType",
+    bound="DaskBackendEntrypoint",
+)
+BackendFuncType = TypeVar("BackendFuncType", bound=Callable[..., Any])
+
+
+class CreationDispatch:
+    """Simple backend dispatch for collection-creation functions"""
+
+    _lookup: dict
+    _module_name: str
+    _config_field: str
+    default: str
+
+    def __init__(self, module_name: str, default: str, name: str | None = None):
+        self._lookup = {}
+        self._module_name = module_name
+        self._config_field = f"{module_name}.backend"
+        self.default = default
+        if name:
+            self.__name__ = name
+
+    def register_backend(
+        self, name: str, backend: BackendEntrypointType
+    ) -> BackendEntrypointType:
+        """Register a target class for a specific backend label"""
+        raise NotImplementedError
+
+    def dispatch(self, backend: str):
+        """Return the desired backend entrypoint"""
+        try:
+            impl = self._lookup[backend]
+        except KeyError:
+            # Check entrypoints for the specified backend
+            entrypoints = detect_entrypoints(f"dask.{self._module_name}.backends")
+            if backend in entrypoints:
+                return self.register_backend(backend, entrypoints[backend].load()())
+        else:
+            return impl
+        raise ValueError(f"No backend dispatch registered for {backend}")
+
+    @property
+    def backend(self) -> str:
+        """Return the desired collection backend"""
+        return config.get(self._config_field, self.default) or self.default
+
+    @backend.setter
+    def backend(self, value: str):
+        raise RuntimeError(
+            f"Set the backend by configuring the {self._config_field} option"
+        )
+
+    def register_inplace(
+        self,
+        backend: str,
+        name: str | None = None,
+    ) -> Callable:
```

Comment:
I think the untyped `Callable` will break the type inference for the functions that this wraps. It may be helpful to put in the work to type the decorator using `ParamSpec`s

Reply:
So, to be a bit more concrete, the following patch types the decorator so that checkers know that the decorator doesn't change the signature of the function:

```diff
diff --git a/dask/backends.py b/dask/backends.py
index 2a5f98543..b478d6b88 100644
--- a/dask/backends.py
+++ b/dask/backends.py
@@ -2,7 +2,9 @@ from __future__ import annotations
 
 from abc import ABC, abstractmethod
 from functools import lru_cache, wraps
-from typing import Any, Callable, TypeVar, cast
+from typing import Any, Callable, TypeVar
+
+from typing_extensions import ParamSpec, TypeAlias
 
 from dask import config
 from dask.compatibility import entry_points
@@ -40,7 +42,8 @@ BackendEntrypointType = TypeVar(
     "BackendEntrypointType",
     bound="DaskBackendEntrypoint",
 )
-BackendFuncType = TypeVar("BackendFuncType", bound=Callable[..., Any])
+BackendFuncParams = ParamSpec("BackendFuncParams")
+BackendFuncReturn = TypeVar("BackendFuncReturn")
 
 
 class CreationDispatch:
@@ -93,10 +96,15 @@ class CreationDispatch:
         self,
         backend: str,
         name: str | None = None,
-    ) -> Callable:
+    ) -> Callable[
+        [Callable[BackendFuncParams, BackendFuncReturn]],
+        Callable[BackendFuncParams, BackendFuncReturn],
+    ]:
         """Register dispatchable function"""
 
-        def decorator(fn: BackendFuncType) -> BackendFuncType:
+        def decorator(
+            fn: Callable[BackendFuncParams, BackendFuncReturn]
+        ) -> Callable[BackendFuncParams, BackendFuncReturn]:
             dispatch_name = name or fn.__name__
             dispatcher = self.dispatch(backend)
             dispatcher.__setattr__(dispatch_name, fn)
@@ -106,7 +114,7 @@ class CreationDispatch:
                 return getattr(self, dispatch_name)(*args, **kwargs)
 
             wrapper.__name__ = dispatch_name
-            return cast(BackendFuncType, wrapper)
+            return wrapper
 
         return decorator
```

The result of this is that we get better inference for, e.g., `read_csv`. So if we look at the following snippet:

```python
from dask.dataframe import read_csv

reveal_type(read_csv)
```

this branch produces
```bash
  /home/ian/dask/dask/typ.py:3:13 - information: Type of "read_csv" is "Unknown"
```
but with the patch we get
```bash
  /home/ian/dask/dask/typ.py:3:13 - information: Type of "read_csv" is "(urlpath: Unknown, blocksize: Unknown = ..., lineterminator: Unknown = ..., compression: Unknown = ..., sample: Unknown = ..., sample_rows: Unknown = ..., enforce: Unknown = ..., assume_missing: Unknown = ..., storage_options: Unknown = ..., include_path_column: Unknown = ..., **kwargs: Unknown) -> Unknown"
```

 which is much closer to correct. It also removes the need for the `cast`.

Reply:
This works great for me locally, but the `typing_extensions` import is problematic for the mindeps check.

Reply:
Hm, forgot that it wasn't installed here yet (though we've almost added it a few times, cf #8674). From my perspective, there's no cost to adding it. It's quite common in the stack at this point. 

In `distributed`, the `typing_extensions` import is guarded by `TYPE_CHECKING`, which would probably work here if we don't want to add the dep.

Perhaps @jrbourbeau or @wence- have thoughts?

Reply:
I do think that guarding the `ParamSpec` import and variable definition with `TYPE_CHECKING` would resolve the problem

Reply:
Okay - I rolled back the change for now, because type-checking takes me longer than I'd like to admit to fumble with. I can improve the behavior here (or in a follow-up) when it is clear what we should do.